### PR TITLE
fix kyua on illumos

### PR DESCRIPTION
**Pull request by ghost**
Related issue: https://github.com/jmmv/kyua/issues/211

struct dirent on solarish (illumos) systems defines d_name member to be char[1] unconditionally (no defines like `__EXTENSIONS__` or the like change that), and readdir_r man page explicitly tells consumers to allocate required storage, so we should do just that.

I'm porting our test suites over to kyua/atf at $work on illumos fork, and while that isn't really FreeBSD related, the original repository does not seem to get any attention anymore, so I have switched over to FreeBSD's forks as the source of truth.

### CODE REVIEW

File: utils/fs/directory.cpp [hunk 1]
```diff
@@ -133,6 +140,10 @@ struct utils::fs::detail::directory_iterator::impl : utils::noncopyable {
     /// readdir_r(3) function.
     ::dirent _dirent;
 
+#if !defined(HAVE_DIRENT_DNAME)
+    char _d_name[NAME_MAX];
```
Comment (kevans91) [inline utils/fs/directory.cpp]: Let's go ahead and throw this `char[NAME_MAX + sizeof(::dirent)]` and the preceding _dirent into an anonymous union to be safe; as noted in IRC, the compiler is allowed to do stupid optimizations with _dirent.d_name based on the apparent size in the current arrangement.

**Comment by kevans91** [comment]
I would have perhaps kept the config glue and just made the new member contingent on that option to avoid the extra `NAME_MAX` reservation on platforms that don't need it, but I don't feel that strongly about it

**Comment by ghost** [comment]
> I would have perhaps kept the config glue and just made the new member contingent on that option to avoid the extra `NAME_MAX` reservation on platforms that don't need it, but I don't feel that strongly about it

I probably will, just wanted to see if checks are OK with this version, somehow I get failures locally.

**Comment by ghost** [comment]
> > I would have perhaps kept the config glue and just made the new member contingent on that option to avoid the extra `NAME_MAX` reservation on platforms that don't need it, but I don't feel that strongly about it
> 
> I probably will, just wanted to see if checks are OK with this version, somehow I get failures locally.

Or, given that checks (at least the one that could pass) are good, and this works for me on illumos, may be we could use this unconditionally?

**Comment by lwhsu** [comment]
Merged as fbb6b99591177f5fb07d3e9f639e5cc58c830d7e

**Pull request closed without merging**